### PR TITLE
New group properties for EWR

### PR DESCRIPTION
**Pull request by emrahkocaman**
1st part of https://github.com/hazelcast/hazelcast/pull/5079

### CODE REVIEW

**Comment by jerrinot** [comment]
could you please add a description + probably a warning properties doesn't have any effect unless Hazelcast Enterprise is used?

**Comment by pveentjer** [comment]
+1. 

Public API needs documentation and the properties are public API.

**Comment by emrahkocaman** [comment]
Javadoc added.

**Comment by jerrinot** [comment]
:+1:

File: hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java [hunk 1]
```diff
@@ -203,7 +203,30 @@
...
+     * Used in case queued event count does not reach to {@link this#PROP_ENTERPRISE_WAN_REP_BATCHSIZE}
+     * Only valid for Hazelcast Enterprise
+     */
+    public static final String PROP_ENTERPRISE_WAN_REP_BATCH_FREQUENCY = "hazelcast.enterprise.wanrep.batchfrequency";
+
+    /**
+     * Defines timeout duration (in milliseconds) for a WAN replication event to reach target cluster,
+     * only valid for Hazelcast Enterprise
+     */
+    public static final String PROP_ENTERPRISE_WAN_REP_OP_TIMEOUT = "hazelcast.enterprise.wanrep.optimeout";
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java]: Please add the unit, e.g. PROP_EVENT_QUEUE_TIMEOUT_MILLIS. Also include it in the string: hazelcast.enterprise.wanrep.optimeout.millis.

Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java]: Sure, done.

File: hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java [hunk 2]
```diff
@@ -203,7 +203,31 @@
...
     public static final String PROP_MAP_WRITE_BEHIND_QUEUE_CAPACITY = "hazelcast.map.write.behind.queue.capacity";
+
+    /**
+     * Defines event queue size for WAN replication, only valid for Hazelcast Enterprise
+     */
     public static final String PROP_ENTERPRISE_WAN_REP_QUEUESIZE = "hazelcast.enterprise.wanrep.queuesize";
+
+    /**
+     * Defines maximum number of WAN replication events to be gathered in a batch, only valid for Hazelcast Enterprise
+     */
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java]: What happens when the batch size is reached? I guess the batch is send. 

What happens when a batch is sending and the new batch grows quickly... will it wait till the previous batch is completed and potentially growing beyond its batch size (potentially leading to oome)? Or will multiple batches be send in parallel? Is the system going to apply back pressure?

File: hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java [hunk 3]
```diff
@@ -203,7 +203,31 @@
...
+    /**
+     * Defines batch sending frequency in seconds,
+     * Used in case queued event count does not reach to {@link this#PROP_ENTERPRISE_WAN_REP_BATCHSIZE}
+     * Only valid for Hazelcast Enterprise
+     */
+    public static final String PROP_ENTERPRISE_WAN_REP_BATCH_FREQUENCY_SECONDS
+            = "hazelcast.enterprise.wanrep.batchfrequency.seconds";
+
+    /**
+     * Defines timeout duration (in milliseconds) for a WAN replication event to reach target cluster,
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java]: The documentation is still insufficient to explain to me what it does. Please look through the eyes of the end user when writing documentation.

File: hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java [hunk 4]
```diff
@@ -203,7 +203,31 @@
     public static final String PROP_ELASTIC_MEMORY_UNSAFE_ENABLED = "hazelcast.elastic.memory.unsafe.enabled";
     public static final String PROP_ENTERPRISE_LICENSE_KEY = "hazelcast.enterprise.license.key";
     public static final String PROP_MAP_WRITE_BEHIND_QUEUE_CAPACITY = "hazelcast.map.write.behind.queue.capacity";
+
+    /**
+     * Defines event queue size for WAN replication, only valid for Hazelcast Enterprise
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java]: What does event queue size mean apart from the size of the event queue? Please provide insight in how fits into the picture and what is being influenced if the property is changed.

File: hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java (hunk 2 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java]: Javadoc updated.

File: hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java (hunk 3 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java]: Javadoc updated.

File: hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java (hunk 4 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/instance/GroupProperties.java]: Javadoc updated.

**Comment by mesutcelik** [comment]
verify

**Comment by mesutcelik** [comment]
:+1: 
seems much better. 

@Serdaro ,
Can you please review javadocs after the merge?

**Comment by Serdaro** [comment]
@david-strom-hazelcast , as we talked last week, could you please copyedit the new content after it is merged?

**Pull request merged**